finetuner/tuner/pytorch/miner.py

```diff
@@ -34,11 +49,61 @@ def mine(
         :return: triplet of label indices follows the order of anchor, positive and negative.
         """
         assert len(embeddings) == len(labels)
-        labels1 = np.expand_dims(labels, 1)
-        labels2 = np.expand_dims(labels, 0)
-        matches = (labels1 == labels2).astype(int)
-        diffs = matches ^ 1
-        np.fill_diagonal(matches, 0)
-        triplets = np.expand_dims(matches, 2) * np.expand_dims(diffs, 1)
-        idxes_anchor, idxes_pos, idxes_neg = np.where(triplets)
-        return list(zip(idxes_anchor, idxes_pos, idxes_neg))
+        return _generate_all_possible_triplets(labels)
+
+
+class SiameseSessionMiner(BaseMiner):
+    def mine(
+        self, embeddings: List[AnyTensor], labels: List[Tuple[int, int]]
+    ) -> List[Tuple[int, ...]]:
+        """Generate tuples from input embeddings and labels.
+
+        :param embeddings: embeddings from model, should be a list of Tensor objects.
+        :param labels: labels of each embeddings, consist of session id and label.
+          the labels are either -1 or 1.
+        :return: a pair of label indices and their label as tuple.
+        """
+        assert len(embeddings) == len(labels)
+        rv = []
+        labels_with_index = [item + (index,) for index, item in enumerate(labels)]
+        sorted_labels_with_index = sorted(labels_with_index, key=lambda x: x[0])
+        for session, group in groupby(sorted_labels_with_index, lambda x: x[0]):
+            _, session_labels, session_indices = zip(*group)
+            pairs = _generate_all_possible_pairs(session_labels)
+            rv.extend(
+                [
+                    (session_indices[left_idx], session_indices[right_idx], label)
+                    for left_idx, right_idx, label in pairs
+                ]
+            )
+        return rv
+
+
+class TripletSessionMiner(BaseMiner):
+    def mine(
+        self, embeddings: List[AnyTensor], labels: List[Tuple[int, int]]
+    ) -> List[Tuple[int, ...]]:
+        """Generate triplets from input embeddings and labels.
+
+        :param embeddings: embeddings from model, should be a list of Tensor objects.
+        :param labels: labels of each embeddings, consist of session id and label.
+        :return: triplet of label indices follows the order of anchor, positive and negative.
+        """
+        assert len(embeddings) == len(labels)
+        rv = []
+        labels_with_index = [item + (index,) for index, item in enumerate(labels)]
+        sorted_labels_with_index = sorted(labels_with_index, key=lambda x: x[0])
+        for session, group in groupby(sorted_labels_with_index, lambda x: x[0]):
+            _, session_labels, session_indices = zip(*group)
+            triplets = _generate_all_possible_triplets(session_labels)
```

Comment:
This implementation is actually wrong. You can not assume that items with label `-1` are similar among each other. I can also see that from your test, because for session `3` you should also get 0 triplets as it is now, but you say you get 6

Reply:
i see, you're right